services/socks5/service/service.go

```diff
@@ -0,0 +1,252 @@
+type NATPinger interface {
+	BindServicePort(key string, port int)
+	Stop()
+	Valid() bool
+}
+
+// NATEventGetter allows us to fetch the last known NAT event
+type NATEventGetter interface {
+	LastEvent() *event.Event
+}
+
+// NewManager creates new instance of SOCKS5 service
+func NewManager(
+	options Options,
+	location location.ServiceLocationInfo,
+	natPort func(port int) (natRelease func()),
+	natEventGetter NATEventGetter,
+	natPinger NATPinger,
+) *Manager {
+	return &Manager{
+		natPort:        natPort,
+		natPinger:      natPinger,
+		natEventGetter: natEventGetter,
+		natPingerPorts: port.NewPool(),
+
+		servicePort: options.Port.Num(),
+		publicIP:    location.PubIP,
+		outboundIP:  location.OutIP,
+	}
+}
+
+// Manager represents an instance of SOCKS5 service
+type Manager struct {
+	natPort        func(int) (natPortRelease func())
+	natPinger      NATPinger
+	natPingerPorts port.ServicePortSupplier
+	natEventGetter NATEventGetter
+
+	servicePort int
+	publicIP    string
+	outboundIP  string
+
+	listener     net.Listener
+	listenerLock sync.Mutex
+	listenerStop chan struct{}
+}
+
+// ProvideConfig provides the config for consumer and handles new SOCKS5 connection.
+func (m *Manager) ProvideConfig(sessionRequestData json.RawMessage) (*session.ConfigParams, error) {
+	if m.servicePort == 0 {
+		return nil, errors.New("Service port not initialized")
+	}
+
+	sessionConfig := sock5_session.Response{
+		Port: m.servicePort,
+	}
+	sessionDestroy := func() {
+		// Do nothing
+	}
+
+	sessionRequest := sock5_session.Request{}
+	err := json.Unmarshal(sessionRequestData, &sessionRequest)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not unmarshal SOCKS5 session request")
+	}
+
+	natPingerEnabled := m.natPinger.Valid() && m.isBehindNAT() && m.portMappingFailed()
+	traversalParams, err := m.newTraversalParams(natPingerEnabled, sessionRequest)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create traversal params")
+	}
+
+	if natPingerEnabled {
+		log.Info().Msgf("NAT Pinger enabled, binding service port for proxy, key: %s", traversalParams.ProxyPortMappingKey)
+		m.natPinger.BindServicePort(traversalParams.ProxyPortMappingKey, sessionConfig.Port)
+		sessionConfig.PortNat = traversalParams.ConsumerPort
+		sessionConfig.Port = traversalParams.ProviderPort
+	}
+
+	return &session.ConfigParams{SessionServiceConfig: sessionConfig, SessionDestroyCallback: sessionDestroy, TraversalParams: &traversalParams}, nil
+}
+
+// Serve starts service - does block
+func (m *Manager) Serve(providerID identity.Identity) error {
+	releasePorts := m.natPort(m.servicePort)
+	defer releasePorts()
+
+	conf := &socks5_lib.Config{}
+	server, err := socks5_lib.New(conf)
+	if err != nil {
+		return errors.Wrap(err, "could not initiate SOCKS5 proxy server")
+	}
+
+	log.Info().Msg("SOCKS5 service started successfully now")
+	err = m.runServer(server, fmt.Sprintf(":%d", m.servicePort))
+	if err != nil && err != errServerStopped {
```

Comment:
Not sure if errServerStopped is needed here. If server is stoped gracefully could return nil and this extra check will not be needed too.